### PR TITLE
Metabuild (RFC 2196)

### PR DESCRIPTION
**Pull request by ehuss**

### CODE REVIEW

**Comment by ehuss** [comment]
There is a major hack/problem with the current implementation and I wanted to see if anyone had ideas on how to fix it.  It currently puts the generated metabuild script into the "target" directory (`target/.metabuild/metabuild-pkg-hash.rs` specifically).  However, the "target" dir is unknown when the `Target` is created.  That means that it doesn't know what to put as the `src_path` in the `Target`.  The current implementation has a hack that will detect the situation later on and use the correct path.  I'd like a more elegant solution, but I haven't been able to come up with anything.  Currently `cargo metadata` displays the wrong path (and that may be quite tricky to fix).

One alternative is to place the `metabuild.rs` script in the root of the package.  However, I'm uncertain about the possible drawbacks of that approach.

The only other idea I had was to try to rewrite the Target list in the Manifest after the "target" directory is known, but I think that will be difficult due to mutability.

Other notes/questions:
- Should there be a dedicated namespace in `Cargo.toml` for metabuild data?  If we use `package.metadata`, then keys may clash with other non-metabuild things.
- Is it OK that it doesn't work with [rename-dependency](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#rename-dependency)?  I'm unable to think of a scenario where that would be needed.
- I'm not a big fan of toml keys that accept multiple types (string or array in this case), since as a new user they tend to be confusing to me (and it only saves 2 characters).  However, I don't feel strongly about it and I implemented it as defined in the RFC.

File: src/cargo/core/compiler/custom_build.rs [hunk 1]
```diff
@@ -584,7 +626,8 @@ pub fn build_map<'b, 'cfg>(cx: &mut Context<'b, 'cfg>, units: &[Unit<'b>]) -> Ca
         }
 
         {
-            let key = unit.pkg
+            let key = unit
+                .pkg
```
Comment (joshtriplett) [inline src/cargo/core/compiler/custom_build.rs]: This seems like an unrelated formatting change that shouldn't appear in the same commit.

File: src/cargo/util/toml/targets.rs [hunk 2]
```diff
@@ -10,21 +10,24 @@
 //! It is a bit tricky because we need match explicit information from `Cargo.toml`
 //! with implicit info in directory layout.
 
-use std::path::{Path, PathBuf};
-use std::fs::{self, DirEntry};
 use std::collections::HashSet;
+use std::fs::{self, DirEntry};
+use std::path::{Path, PathBuf};
```
Comment (joshtriplett) [inline src/cargo/util/toml/targets.rs]: This seems like an unrelated formatting change. (It's a good change, but unrelated to this commit.)

File: src/cargo/util/toml/targets.rs [hunk 3]
```diff
@@ -10,21 +10,24 @@
...
 
+use super::{
+    LibKind, PathValue, StringOrBool, StringOrVec, TomlBenchTarget, TomlBinTarget,
+    TomlExampleTarget, TomlLibTarget, TomlManifest, TomlTarget, TomlTestTarget,
+};
 use core::{compiler, Edition, Target};
 use util::errors::CargoResult;
-use super::{LibKind, PathValue, StringOrBool, TomlBenchTarget, TomlBinTarget, TomlExampleTarget,
-            TomlLibTarget, TomlManifest, TomlTarget, TomlTestTarget};
```
Comment (joshtriplett) [inline src/cargo/util/toml/targets.rs]: Likewise, unrelated format change.

File: src/cargo/util/toml/targets.rs [hunk 4]
```diff
@@ -476,7 +500,14 @@ fn clean_targets_with_legacy_path(
...
     for target in toml_targets {
-        let path = target_path(&target, inferred, target_kind, package_root, edition, legacy_path);
+        let path = target_path(
+            &target,
+            inferred,
+            target_kind,
+            package_root,
+            edition,
+            legacy_path,
+        );
```
Comment (joshtriplett) [inline src/cargo/util/toml/targets.rs]: Unrelated format change.

File: src/cargo/core/compiler/custom_build.rs [hunk 5]
```diff
@@ -536,6 +540,44 @@ impl BuildOutput {
...
+    // Filter out optional dependencies, and look up the actual lib name.
+    let meta_deps: Vec<_> = deps
+        .iter()
+        .filter_map(|name| {
+            available_deps
+                .iter()
+                .find(|u| u.pkg.name().as_str() == name.as_str())
+                .map(|dep| dep.target.crate_name())
+        })
+        .collect();
```
Comment (joshtriplett) [inline src/cargo/core/compiler/custom_build.rs]: As a minor nit that I'm not sure needs fixing: this is O(n^2), because it searches every available_deps for every metadep. If some readily available data structure provides an O(1) mapping from name to dep, I'd suggest using that. If one doesn't exist, though, I don't know that it's worth creating one just for this, given that the common case will have relatively few metabuild entries.

File: src/cargo/core/compiler/custom_build.rs [hunk 6]
```diff
@@ -536,6 +540,44 @@ impl BuildOutput {
...
+    for dep in &meta_deps {
+        output.push(format!("extern crate {};\n", dep));
+    }
+    output.push("fn main() {\n".to_string());
+    for dep in &meta_deps {
+        output.push(format!("    {}::metabuild();\n", dep));
+    }
+    output.push("}".to_string());
```
Comment (joshtriplett) [inline src/cargo/core/compiler/custom_build.rs]: Minor nit: please include a newline at the end of the file.

File: src/cargo/core/compiler/custom_build.rs [hunk 7]
```diff
@@ -536,6 +540,44 @@ impl BuildOutput {
...
+    let output = output.join("");
+    let path = cx.files().metabuild_path(unit);
+    let changed = if let Ok(orig) = fs::read_to_string(&path) {
+        orig != output
+    } else {
+        true
+    };
+    if changed {
+        fs::write(&path, output)?;
+    }
```
Comment (joshtriplett) [inline src/cargo/core/compiler/custom_build.rs]: For clarity and potential future reuse, please extract this logic into a "write if changed" function.

**Comment by joshtriplett** [comment]
Thank you for working on this, @ehuss!

File: src/cargo/core/compiler/custom_build.rs (hunk 1 above)
Comment (ehuss) [reply src/cargo/core/compiler/custom_build.rs]: Yea, I prefer to use rustfmt to format, but master has diverged quite a bit from the canonical format.  I'll try to remove the unrelated changes.

I was thinking I might post a separate PR running rustfmt, but that touches 123 files which I think probably can't be reviewed properly.  I'm not sure how you guys want to handle that moving forward.  Each rust release brings more changes to rustfmt.  It makes things more difficult the more the master branch diverges.

**Comment by joshtriplett** [comment]
Other than the issue you mentioned with the target directory, this looks good to me now.

**Comment by alexcrichton** [comment]
Egad I had missed this until now! @ehuss is this ready to go after a once over or were there still any remaining design questions we wanted to answer before landing?

**Comment by ehuss** [comment]
There's one major problem left.  The `Target::src_path` needs to be set to point into the `target` directory for the metabuild script.  However, the `target` directory is not known until after all the manifests are parsed.  Currently there is [a hack](https://github.com/rust-lang/cargo/pull/5628/files#diff-7f98585dbf9d30aa100c8318e2c77e79R644) to adjust the path, but the `cargo metadata` output is still wrong.  I haven't come up with any solutions that don't involve a huge number of changes.  The method I was leaning towards is modifying the `src_path` after parsing is done, but that would involve keeping the manifests `mut`, and that just sounded really unpleasant.  But I haven't tried, yet.  I was kinda hoping someone had a simpler idea. 😄

**Comment by alexcrichton** [comment]
Ok sure yeah makes sense. I think though that the `cargo metadata` output is just gonna need to change somewhat significantly to cope with this. We don't really have a source path to write down in the sense that `cargo metadata` probably shouldn't be generating files anyway?

**Comment by ehuss** [comment]
Would it be OK if `cargo metadata` showed the path to the generated metabuild script, even if that file might not exist, yet?  (i.e. the file is only created if you run `cargo build`.)

Example:
```json
{"packages": [
    {"name": "foo",
     "targets": [
        {"kind": ["custom-build"],
         "crate_types": ["bin"],
         "name": "foo",
         "src_path": ["/path/to/foo/target/.metabuild/metabuild-foo-e8a1568fc72dd376.rs"]
        }, 
        ...
     ],
     ...
    }
]}
```

**Comment by alexcrichton** [comment]
Hm perhaps yeah, I was thinking "no" but I'm also thinking more of build plans. It probably wouldn't be the end of the world for sure! In any case this is all unstable so as long as it's mentioned on the tracking issue it's fine to implement whatever for now I think

**Comment by ehuss** [comment]
I hacked together an alternate solution that mutates the `src_path` after the workspace is loaded.  You can see it here: https://github.com/ehuss/cargo/commit/a225f5f3ecfe4c2cf637f855bb63d12cc30b1625  In particular, note the `fixup()` function.  I'm not so certain if this is better or not.  This new version uses a different path (`target/.metabuild/metabuild-NAME-HASH.rs`) since the release doesn't affect the file's contents. It also returns a real path in `cargo metadata`.  I also added a test to illustrate that `cargo build --build-plan` should do the right thing (the generated build script is created), so I think it can still be compatible with external build systems.

The only other approach I've thought of is to change `Target::src_path` to an Option or some other enum, and then do something different in `cargo metadata`, but that seems worse to me.

Let me know which approach sounds the most palatable, or if you have some other idea.  This new one works a little better (and is a little simpler), so I'm inclined towards that.

**Comment by ehuss** [comment]
> (which is what I think we want the end goal to be?)

Hmm, I'm not sure.  At least for build-plan, you'd want it to be there, otherwise you wouldn't be able to use this feature with external build systems, right?

As for `cargo metadata`, what's the reasoning for hiding it?  Would it just skip serializing the entire target?  (I don't think we can just skip the field or use null, since it would be a breaking change?)  I'm not completely familiar with all the use cases.  I have a tool that uses `src_path`, but it would ignore these paths.  Are you thinking of tools that inspect these files, and would fail if the file doesn't exist?

**Comment by alexcrichton** [comment]
I'm not sure actually that this would necessarily be compatible with a build plan! In some contexts you may move around where execution happens and the original generation of the plan may happen in a somewhat isolated environment which isn't guaranteed to have changes reflected into others per se. I sort of see this as the responsibility of the build planner to figure out how to assemble this file to compile it, although it's true that the source is always generated by cargo so maybe it is better to include it then?

I think in general I'm just worried about build plans and `cargo metadata` writing into the `target` directory when they're in theory read-only commands. That may cause some issues down the road maybe, but I'm not entirely sure.

In any case this is all unstable so I'm fine landing whatever here for now! Just want to make sure we keep tabs on this in the tracking issue

**Comment by ehuss** [comment]
> my preference would be to reflect this in Target itself and just store Option<PathBuf> (effectively) for now.

I've been trying to figure out how to do this.  I'm not sure I understood this suggestion.  `src_path` can't change to an Option since that would be a breaking change for the JSON output, correct?  I could filter out the `Target` from the list in `cargo metadata`, but `Target` is serialized in other places (like JSON compiler messages) where there is no alternative, so that might not help.

**Comment by alexcrichton** [comment]
It's a breaking change I think yeah in the sense that consumers who aren't ready for this type of package won't work, but that's also somewhat expected when we add new features to Cargo in that older clients aren't always compatible with newer features. It doesn't break any existing build, though, which I believe is the most important part.

**Comment by joshtriplett** [comment]
cc @wycats @aturon

I'd like to get this unstuck.

**Comment by ehuss** [comment]
Sorry for letting this languish.  I would prefer not to break the JSON output, as it would break a lot of tools.  I agree it is not perfect to expose the path to a mostly internal file/implementation detail.  However, exposing it seems to be less harmful than changing the output.  The only concern I can think of is if someone runs `cargo metadata` before any other command, and expects every source file to exist, it will have a missing file.

Also, for the build-plan portion, I'm not convinced hiding it is a good idea.  I don't see how an external build tool would be able to construct the build script and build it properly without the metabuild included in the build-plan output.  I think it would essentially make this unusable with external build tools.

**Comment by alexcrichton** [comment]
I basically have no opinion on how this should be implemented, I'm totally fine with basically anything because it's unstable. Just need to be sure to note it down to discuss before stabilization!

**Comment by ehuss** [comment]
I have updated with a new approach of using a custom enum for `Target::src_path`.  I'm not entirely happy with this strategy, because it is a little brittle (particularly around serialization).  However, it works better than other approaches I've tried, and avoids changing the JSON structures.  In particular:

* `cargo metadata` now *skips* the metabuild target.
* Artifact JSON messages include the true path.  I'm not familiar with the use cases for these messages, so I'm not sure if that's really the best approach.
* Build plans continue to include the information to build the metabuild script.

I think this is now ready for review, I'm not aware of any remaining issues.

**Comment by alexcrichton** [comment]
Thanks!
